packages/flutter_test/lib/src/controller.dart

```diff
@@ -31,12 +226,33 @@ const String _defaultPlatform = kIsWeb ? 'web' : 'android';
 ///
 /// Concrete subclasses must implement the [pump] method.
 abstract class WidgetController {
-  /// Creates a widget controller that uses the given binding.
-  WidgetController(this.binding);
+  /// Creates a widget controller that uses the given binding. A [SemanticsController]
+  /// will be created for [WidgetController.semantics] by default, but a custom
+  /// implementation can be provided for advanced use cases.
+  WidgetController(this.binding, [SemanticsController? semantics])
```

Comment:
Whats the use case where people would want to create one?

Reply:
TBH, I'm not 100% sure on a use case for a custom `SemanticsController` implementation. I have a gut feeling that integration testing may need a different implementation when we start adding to `SemanticsController` for more advanced scenarios like #112413.

The binding that's passed to `SemanticsController` may take care of those more advanced scenarios, but I'm not certain how well it covers semantics. That combined with the recent request in #113097 made me think that it would be better to be flexible here instead of having to refactor later.

@goderbauer mentioned this as well, and I'm willing to remove the ability to pass a custom SemanticsController for the moment if either of you request it, I just have a sneaking suspicion it'll be useful.

Reply:
I don't think any of the platform can do reasonable a11y integration test other than checking semantics attribute on semantics node, which will be the same thing in unit test. I would suggest we hide it for now, we can always expose it later if needs to

Reply:
Fair enough, I've removed the optional `semantics` parameter from `WidgetController`, its' implementations, and the `testWidgets` method.